testnet/stacks-node/src/nakamoto_node/relayer.rs

```diff
@@ -641,6 +661,71 @@ impl RelayerThread {
         Ok(())
     }
 
+    fn continue_tenure(&mut self, new_burn_view: ConsensusHash) -> Result<(), NakamotoNodeError> {
+        if let Err(e) = self.stop_tenure() {
+            error!("Relayer: Failed to stop tenure: {:?}", e);
+            return Ok(());
+        }
+        debug!("Relayer: successfully stopped tenure.");
+        // Check if we should undergo a tenure change to switch to the new burn view
+        let burn_tip =
+            SortitionDB::get_block_snapshot_consensus(self.sortdb.conn(), &new_burn_view)
+                .map_err(|e| {
+                    error!("Relayer: failed to get block snapshot for new burn view: {e:?}");
+                    NakamotoNodeError::SnapshotNotFoundForChainTip
+                })?
+                .ok_or_else(|| {
+                    error!("Relayer: failed to get block snapshot for new burn view");
+                    NakamotoNodeError::SnapshotNotFoundForChainTip
+                })?;
+
+        let (canonical_stacks_tip_ch, canonical_stacks_tip_bh) =
+            SortitionDB::get_canonical_stacks_chain_tip_hash(self.sortdb.conn()).unwrap();
+        let canonical_stacks_tip =
+            StacksBlockId::new(&canonical_stacks_tip_ch, &canonical_stacks_tip_bh);
+        let block_election_snapshot =
+            SortitionDB::get_block_snapshot_consensus(self.sortdb.conn(), &canonical_stacks_tip_ch)
+                .map_err(|e| {
+                    error!("Relayer: failed to get block snapshot for canonical tip: {e:?}");
+                    NakamotoNodeError::SnapshotNotFoundForChainTip
+                })?
+                .ok_or_else(|| {
+                    error!("Relayer: failed to get block snapshot for canonical tip");
+                    NakamotoNodeError::SnapshotNotFoundForChainTip
+                })?;
+
+        let Some(ref mining_key) = self.config.miner.mining_key else {
+            return Ok(());
+        };
+        let mining_pkh = Hash160::from_node_public_key(&StacksPublicKey::from_private(mining_key));
+        if block_election_snapshot.miner_pk_hash != Some(mining_pkh) {
```

Comment:
`block_election_snapshot` would always resolve to the snapshot of the highest-processed Stacks tenure, not the last snapshot with a sortition, right?  The sortition DB can advance without the Stacks chain tip advancing (this happens all the time, since the first Stacks block for a given burn view consensus hash will often arrive after that burn view's sortition).  If so, then I think there might be a bug here.  You need to be comparing the miner key of the snapshot with the last non-empty sortition on the canonical burnchain fork to the ongoing miner.  You can get this snapshot with `SortitionHandleConn::get_last_snapshot_with_sortition()`.  If the miner key in this snapshot is the same as the node's mining key, then the node can issue a tenure extension (note that this also covers the case where there is one or more empty sortitions after this last-sortition snapshot)  Otherwise, either this node won the last sortition, or someone else did.  If someone else did, then this node stops mining.  But if this node won sortition again, I think it actually needs to issue a `BlockFound` tenure change because that's what earns it a coinbase.

Reply:
I think the case of BlockFound is already handled in `process_sortition`. We check in that function whether there was a sorition in the last snapshot and if we have won it. If both are true, we issue a block found. Otherwise if no sortition existed, we attempt a continue tenure (which will only occur in the case of us having the same key and therefore were the winner of the last won sortition)

Reply:
Interestingly I am having a hard time. If I do get_last_snapshot_with_sortition instead of my current get_block_snapshot_consensus, my tests hang. Still looking into.